From 4c33b1cb6400341121700473da03d4083990a093 Mon Sep 17 00:00:00 2001
From: Claudiu Zissulescu <claziss@gmail.com>
Date: Mon, 1 Apr 2019 13:03:55 +0300
Subject: [PATCH 2/7] [ARC] Fix for PR89838.

gcc/
xxxx-xx-xx  Claudiu Zissulescu  <claziss@synopsys.com>

	* config/arc/arc.c (arc_legitimize_pic_address): Emit an move
	instruction if needed.
	(prepare_pic_move): Always provide an spare register.

/gcc/testsuite
xxxx-xx-xx  Claudiu Zissulescu  <claziss@synopsys.com>

	* gcc.target/arc/pr89838.c: New file.
---
 gcc/config/arc/arc.c                   |  9 ++++++++-
 gcc/testsuite/gcc.target/arc/pr89838.c | 16 ++++++++++++++++
 2 files changed, 24 insertions(+), 1 deletion(-)
 create mode 100644 gcc/testsuite/gcc.target/arc/pr89838.c

diff --git a/gcc/config/arc/arc.c b/gcc/config/arc/arc.c
index ba2e3dd8208..83664d4d64f 100644
--- a/gcc/config/arc/arc.c
+++ b/gcc/config/arc/arc.c
@@ -6093,6 +6093,13 @@ arc_legitimize_pic_address (rtx orig, rtx oldx)
 	  if (base == op0 && pat == op1)
 	    return orig;
 
+	  if (GET_CODE (base) == PLUS)
+	    {
+	      gcc_assert (oldx != NULL_RTX);
+	      gcc_assert (REG_P (oldx));
+	      emit_insn (gen_rtx_SET (oldx, base));
+	      base = oldx;
+	    }
 	  if (GET_CODE (pat) == CONST_INT)
 	    pat = plus_constant (Pmode, base, INTVAL (pat));
 	  else
@@ -6285,7 +6292,7 @@ prepare_pic_move (rtx *operands, machine_mode)
   else
     {
       rtx temp = (reload_in_progress ? operands[0]
-		  : flag_pic? gen_reg_rtx (Pmode) : NULL_RTX);
+		  : gen_reg_rtx (Pmode));
       operands[1] = arc_legitimize_pic_address (operands[1], temp);
     }
 }
diff --git a/gcc/testsuite/gcc.target/arc/pr89838.c b/gcc/testsuite/gcc.target/arc/pr89838.c
new file mode 100644
index 00000000000..559434ac87e
--- /dev/null
+++ b/gcc/testsuite/gcc.target/arc/pr89838.c
@@ -0,0 +1,16 @@
+/* { dg-do compile } */
+/* { dg-require-effective-target tls } */
+/* { dg-options "-O2" } */
+
+extern void foo (void);
+extern void bar (void *);
+
+struct {
+  int __attribute__(()) a;
+  int __attribute__(()) b;
+} __thread c __attribute__((tls_model("initial-exec")));
+
+void foo (void)
+{
+  bar (&c.b);
+}
-- 
2.16.2

